BrainPortal/app/helpers/userfiles_helper.rb

```diff
@@ -91,27 +92,22 @@ def data_link(file_name, userfile)
     return h(display_name) unless file_lstat.file?
 
     matched_class = SingleFile.descendants.unshift(SingleFile).find { |c| file_name =~ c.file_name_pattern }
-
-    if matched_class && matched_class <= TextFile
-      link_to h(display_name),
-              display_userfile_path(userfile,
-                :content_loader        => :collection_file,
-                :arguments             => file_name,
-                :viewer                => :text_file,
-                :viewer_userfile_class => :TextFile,
-                :content_viewer        => "off",
-              ),
-              :target => "_blank"
-    elsif matched_class && matched_class <= ImageFile
-      link_to h(display_name),
-              display_userfile_path(userfile,
-                :content_loader        => :collection_file,
-                :arguments             => file_name,
-                :viewer                => :image_file,
-                :viewer_userfile_class => :ImageFile,
-                :content_viewer        => "off",
-              ),
-              :target => "_blank"
+    viewer        = matched_class.class_viewers.first.partial rescue nil
```

Comment:
I think searching the viewers list by matching with the partial names is wrong. Viewers shoudl be refernced by name, not but the partial names

Reply:
Of course sometimes the name of the partial is the same as the name of the viewer, but not always.